distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso

```diff
@@ -613,36 +627,52 @@ type Vector a
 
              [1] + [2]
     + : Vector Any -> Vector Any
-    + self that =
-        self_len = self.length
-        arr = Array.new (self_len + that.length)
-        Array.copy self.to_array 0 arr 0 self_len
-        Array.copy that.to_array 0 arr self_len that.length
-        Vector.from_polyglot_array arr
-
-    ## Add `element` to the beginning of `self` vector.
+    + self that = case that of
+        _ : Vector ->
+            self_len = self.length
+            arr = Array.new (self_len + that.length)
+            Array.copy self.to_array 0 arr 0 self_len
+            Array.copy that.to_array 0 arr self_len that.length
+            Vector.from_polyglot_array arr
+        _ : Array -> self + Vector.from_polyglot_array that
+        _ -> Error.throw (Type_Error.Error Vector that "Should be a Vector or Array to be appended.")
+
+    ## Inserts the given item into the vector at the given index.
 
        Arguments:
-       - `element`: An element to add to this vector.
-
-       > Example
-         Add 1 to the start of the vector.
-
-             [2, 3].prepend 1
-    prepend : Any -> Vector Any
-    prepend self element = [element] + self
-
-    ## Add `element` to the end of `self` vector.
+       - at: The index at which to insert the item.
+         If the index is greater than the length of the vector, the item will
+         be added at the end.
+         If the index is less than 0, the index will be counted back from the
+         end. If before the beginning will be added at the start.
+       - item: The item to insert into the vector.
+    insert : Integer -> Any -> Vector
+    insert self at=self.length item=Nothing =
```

Comment:
For `0 <= i <= vec.length` we have an invariant

`(vec.insert i x) . at i == x`

However, for `i > vec.length` this does not hold anymore, i.e. `[1,2,3].insert 100 4` will not insert it at 100th index. Do we surely want that?

I'd lean towards one of the following, if `at > self.length`:
1. throw an error - if one wants to insert at the end, they should use `at=self.length`, which is even a default, if they are putting a greater index it may indicate a bug in their index-computing logic.
2. pad the Vector with nulls, i.e. `[0,1].insert 4 4 == [0, 1, Nothing, Nothing, 4]`, so that the invariant holds. I'm less convinced about this as the padding behaviour may be unexpected too.

So I'd suggest to go for (1) and throw an error if the index is too large. Do you think that would be OK? Or is there some particular reason for appending it at the end and silently ignoring the fact that the index was too large? Is there a practical use case for this that I missed?

<hr>

One additional argument for why I think this should be an error is that it may be tempting to use some large number to ensure the element is always inserted at the end, let's say `vec.insert 100 "X"`. But then if `vec` is longer than `100` it will actually not insert at the end. So I don't think we should 'tempt' the users to writing such incorrect code by allowing this behaviour. Especially, as the default arg is already doing the correct 'insert at end'.

Reply:
Will make it error if outside the range of [0, length] after normalising for negative index.